### PR TITLE
chore: cleanup pg notifier

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Introducing **repository** and **model** for Notifier.
- Renamed it to `notifier` instead of `pgNotifier` since we don't need to expose the implementation details like it is implemented for Postgres.
- Removing globals.
- Removed `notifier.Init()`.
- Addition of `claimLag` and `trackBatchLag` for monitoring purposes.
  - `claimLag` to track between the time when the record was created and the time when it was claimed.
  - `trackBatch` to track the number of track batches running.
  - `trackBatchLag` to track the time and how much is being spent on tracking per publish.
  - `orphanJobs` to observe the number of orphan jobs by maintenance.
- Introduced `background errgroup` to **gracefully shut down the goroutines** being created during **track batch** and **subscribe**.
- Added a separate table called `**pg_notifier_queue_metadata**` for **storing common metadata** instead of sending them in each payload as a separate transaction in an **attempt to make the Notifier Generic**. It will help us create **bigger batches**, and **less tracking** or **polling** for batches to complete. 
- Started using the **topic** to make it **version-compatible while claiming and publishing**. The major use case is, that when the release happens, slaves come first and they start processing jobs and after a while when the master is up, it deletes them. So, it's better to have the compatibility.
- **Moving maintenance workers to Master instead of Slave.** Earlier for the Multi-tenant setup, we used to have many rudder server nodes running in embedded mode and share some shared slaves pool. Now warehouse master is kind of Multi-tenant.

# Discussion
- Should we get rid of maintenance workers by getting a database lock?
- Instead of a claim, can we probably have a `trigger` that updates when the status is marked as failed or waiting?

## Linear Ticket

- https://linear.app/rudderstack/issue/PIPE-169/notifier-cleanup

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Comment by achettyiitr** [comment]
Separate the repo for master and slave.

File: services/notifier/notifier.go [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+	n.config.password = n.conf.GetString("PGNOTIFIER_DB_PASSWORD", "ubuntu")
+	n.config.sslMode = n.conf.GetString("PGNOTIFIER_DB_SSL_MODE", "disable")
+	n.config.maxAttempt = n.conf.GetInt("PgNotifier.maxAttempt", 3)
+	n.config.maxOpenConnections = n.conf.GetInt("PgNotifier.maxOpenConnections", 20)
+	n.config.shouldForceSetLowerVersion = n.conf.GetBool("SQLMigrator.forceSetLowerVersion", true)
+	n.config.trackBatchInterval = n.conf.GetDuration("PgNotifier.trackBatchIntervalInS", 2, time.Second)
+	n.config.queryTimeout = n.conf.GetDuration("Warehouse.pgNotifierQueryTimeout", 5, time.Minute)
+
+	n.conf.RegisterDurationConfigVariable(5000, &n.config.maxPollSleep, true, time.Millisecond, "PgNotifier.maxPollSleep")
+	n.conf.RegisterDurationConfigVariable(120, &n.config.jobOrphanTimeout, true, time.Second, "PgNotifier.jobOrphanTimeout")
```
Comment (achettyiitr) [inline services/notifier/notifier.go:115]: Hot reloadables should be atomic or be read when needed.
https://github.com/rudderlabs/rudder-server/pull/3805#discussion_r1311266977

File: services/notifier/notifier.go [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+		}
+	}
+
+	for {
+		orphanJobIDs, err := n.repo.OrphanJobIDs(ctx, int(n.config.jobOrphanTimeout/time.Second))
+		if err != nil {
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return nil
+			}
+			if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == "57014" {
```
Comment (fracasula) [inline services/notifier/notifier.go:472]: This doesn't take into consideration that the error could have been wrapped and yet be a `*pq.Error` under the hood.

```go
package main

import (
	"errors"
	"fmt"

	"github.com/lib/pq"
)

func main() {
	originalErr := &pq.Error{
		Code:             "123",
		Message:          "Hello World",
	}
	wrappedErr := fmt.Errorf("wrapped: %w", originalErr)

	var pqErr *pq.Error
	if errors.As(wrappedErr, &pqErr) && pqErr.Code == "123" {
		fmt.Println("errors.As: pq error code is 123")
	}

	if pqErr, ok := wrappedErr.(*pq.Error); ok && pqErr.Code == "57014" {
		fmt.Println("typeAssertion: pq error code is 123")
	}
}
```

https://go.dev/play/p/kxL4eZG475K?v=

As you can see `typeAssertion: pq error code is 123` did not get printed, but `errors.As: pq error code is 123` did.

**Review by fracasula — COMMENTED** [review]
@achettyiitr quick preliminary review, I don't think I'll be able to finish the review this week and next week I'm off. I know this one it's not urgent so maybe it can wait a bit or you can get reviews from other people :+1:

**Comment by achettyiitr** [comment]
> @achettyiitr quick preliminary review, I don't think I'll be able to finish the review this week and next week I'm off. I know this one it's not urgent so maybe it can wait a bit or you can get reviews from other people 👍

Yeah.

File: services/notifier/notifier.go [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+	queueName = "pg_notifier_queue"
+	module    = "pgnotifier"
+)
+
+type Notifier struct {
+	conf                *config.Config
+	logger              logger.Logger
+	statsFactory        stats.Stats
+	db                  *sqlmw.DB
+	repo                *repo.Notifier
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply services/notifier/notifier.go]: Good points.

File: services/notifier/notifier.go [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+		maxAttempt                 int
+		maxOpenConnections         int
+		shouldForceSetLowerVersion bool
+		trackBatchInterval         time.Duration
+		maxPollSleep               time.Duration
+		jobOrphanTimeout           time.Duration
+		queryTimeout               time.Duration
+	}
+	stats struct {
+		insertRecords      stats.Measurement
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply services/notifier/notifier.go]: Thanks for pointing out. Was not aware of whether we have stats type-specific interfaces.

File: services/notifier/notifier.go [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+		return n.config.maxPollSleep
+	}
+
+	n.background.group.Go(func() error {
+		defer close(jobsCh)
+
+		pollSleep := time.Duration(0)
+
+		for {
+			job, metadata, err := n.claim(ctx, workerId)
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply services/notifier/notifier.go:380]: Yeah, we can log it if it is not `sql.ErrNoRows` and `context.Canceled`

File: services/notifier/notifier.go [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+		claimSucceededTime stats.Measurement
+		claimFailed        stats.Measurement
+		claimFailedTime    stats.Measurement
+		claimUpdateFailed  stats.Measurement
+		abortedRecords     stats.Measurement
+	}
+}
+
+func New(
+	ctx context.Context,
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply services/notifier/notifier.go:96]: Thanks, @fracasula for pointing it out and explaining it in a detailed way. This is super helpful for me to clearly understand the difference between setup vs. start.

[the remark this answers is not in the document]
Comment (achettyiitr) [reply services/notifier/notifier.go:96]: If we are not sure of the number of background workers, I am currently using Wait() to wait for the background `ctx` to get canceled and wait for all the drained jobs to be completed.
Do you have some opinion on this? I took it from `errgroup` implementation.

File: services/notifier/repo/subscriber.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	query := fmt.Sprintf(`
+		UPDATE
+		  `+notifierTableName+`
+		SET
+		  status =(
+			CASE WHEN attempt > $1 THEN CAST (
+			  '%[1]s' AS pg_notifier_status_type
+			) ELSE CAST(
+			  '%[2]s' AS pg_notifier_status_type
+			) END
```
Comment (lvrach) [inline services/notifier/repo/subscriber.go:82]: For consistency, readability and performance.

```suggestion
			CASE WHEN attempt > $1 THEN CAST (
			  '` + model.Aborted + `' AS pg_notifier_status_type
			) ELSE CAST(
			   '` + model.Failed + `'  AS pg_notifier_status_type
			) END
```

File: services/notifier/repo/subscriber.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+		  id = $4;
+	`,
+		model.Aborted,
+		model.Failed,
+	)
+
+	_, err := n.db.ExecContext(ctx,
+		query,
+		maxAttempt,
+		n.now().Format(timeFormat),
```
Comment (lvrach) [inline services/notifier/repo/subscriber.go:97]: Why do we have to format time here? Would it be an issue if we passed the `time.Time` ?

File: services/notifier/repo/subscriber.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+		misc.QuoteLiteral(claimError.Error()),
```
Comment (lvrach) [inline services/notifier/repo/subscriber.go:98]: Do we have to `QuoteLiteral` given we passed it as a prepared statement

**Review by lvrach — APPROVED** [review]
LGTM, minor comments.

Great job @achettyiitr

File: services/notifier/repo/subscriber.go (hunk 9 above)
Comment (achettyiitr) [reply services/notifier/repo/subscriber.go:98]: Thanks for pointing out.

File: services/notifier/repo/subscriber.go (hunk 8 above)
Comment (achettyiitr) [reply services/notifier/repo/subscriber.go:97]: It's not required. I will get rid of this.

**Pull request closed without merging**